wagtail/admin/filters.py

```diff
@@ -67,6 +68,128 @@ def decompress(self, value):
+                filter_value = None
+
+            if option_value is None:
+                option_value = ''
+
+            subgroup = []
+            if isinstance(option_label, (list, tuple)):
+                group_name = option_value
+                subindex = 0
+                choices = option_label
+            else:
+                group_name = None
+                subindex = None
+                choices = [(option_value, option_label)]
+            groups.append((group_name, subgroup, index))
+
+            for subvalue, sublabel in choices:
+                selected = (
+                    str(subvalue) in value
+                    and (not has_selected or self.allow_multiple_selected)
+                )
+                has_selected |= selected
+
+                subgroup.append(self.create_option(
+                    name, subvalue, sublabel, selected, index, subindex=subindex,
+                    filter_value=filter_value
+                ))
+                if subindex is not None:
+                    subindex += 1
+        return groups
+
+    def create_option(self, name, value, label, selected, index, subindex=None, attrs=None, filter_value=None):
+        option = super().create_option(
+            name, value, label, selected, index, subindex=subindex, attrs=attrs
+        )
+        if filter_value is not None:
+            option['attrs']['data-filter-value'] = filter_value
+
+        return option
+
+    @property
+    def media(self):
+        return forms.Media(js=[
+            versioned_static('wagtailadmin/js/filtered-select.js'),
+        ])
+
+
+class FilteredModelChoiceIterator(django_filters.fields.ModelChoiceIterator):
+    """
+    A variant of Django's ModelChoiceIterator that, instead of yielding (value, label) tuples,
+    returns (value, label, filter_value) so that FilteredSelect can drop filter_value into
+    the data-filter-value attribute.
+    """
+    def choice(self, obj):
+        return (
+            self.field.prepare_value(obj),
+            self.field.label_from_instance(obj),
+            self.field.get_filter_value(obj)
+        )
+
+
+class FilteredModelChoiceField(django_filters.fields.ModelChoiceField):
+    widget = FilteredSelect
+    iterator = FilteredModelChoiceIterator
+
+    def __init__(self, *args, **kwargs):
+        self.filter_accessor = kwargs.pop('filter_accessor')
+        filter_field = kwargs.pop('filter_field')
+        super().__init__(*args, **kwargs)
+        self.widget.filter_field = filter_field
+
+    def get_filter_value(self, obj):
+        # filter_accessor identifies a property or method on the instances being listed here,
+        # which gives us a queryset of related objects. Turn this queryset into a list of IDs
+        # that will become the 'data-filter-value' used to filter this listing
+        queryset = getattr(obj, self.filter_accessor)
+        if callable(queryset):
+            queryset = queryset()
+
```

Comment:
This limits it to methods/attributes of the instance itself, as far as I can see? In this case that works really well, I think it's a good addition to Task, but would it be worth adding a way to specify a queryset/filter/function more generally here, that just takes the instance as an argument? Eg that would let me specify the `Workflow.objects.filter(workflow_tasks__task=instance)`?

Reply:
Makes sense, yep - it wouldn't be a massive leap to allow passing a callable for `filter_accessor`, so that you could write something like:

    task = FilteredModelChoiceFilter(
        queryset=Task.objects.all(),
        filter_field='id_workflow',
        filter_accessor=(lambda task:Workflow.objects.filter(workflow_tasks__task=task))
    )

At this point I think the harder task is going to be documenting what's going on, since this is quickly outgrowing "a bit of supporting code to make this dropdown work" into a reusable component (and, quite possibly, its own mini-library outside of Wagtail...)

Reply:
True, it is growing! I think having it reusable would be great (but yeah, documenting it somewhere does sound good). Possibly the documentation section for reports filtering, once we make that public?

Reply:
Now accepts relations and arbitrary callables as of caa81cc807b4ff65de35038a1bb11d3cc7510ab1.